cpp/leaky_core.cc

```diff
@@ -18,6 +19,11 @@
 
 namespace pyos {
 
+const int kMaxSigs = 128;
+const int kMaxSigTraps = 128;
+builtin_trap::_TrapHandler* signal_traps[kMaxSigs][kMaxSigTraps];
```

Comment:
Hm so why is it a 2 dimensional array?  

I haven't looked  at this code in awhile, but my memory is that the bash trap builtin has two kinds of "sig specs":

- a Unix signal like SIGINT / INT
- a "user space" hook like ERR or DEBUG

I guess right now those are mixed in the "self.traps" dictionary

I think the key weirdness is that we have state in that dictionary, and state in the global array

Let me think about this

Reply:
Ah yes, this is a little awkward. Basically the problem is, the current `_TrapHandler` type has no context for which signal (if not a hook) it was registered to. So, we keep a list of traps for each possible signal. This is pretty inefficient... If we added that missing context we could flatten this array and simplify a lot of the dependent code.

Luckily we don't have this problem for hooks, since the python code for that is portable and mycpp translates it cleanly